content/blog/multi-single-tenant-github-app-installation/index.md

```diff
@@ -0,0 +1,62 @@
+- The admin user on-boarding onto a newly created roadie backstage instance must be hinted to install the “Roadie Backstage” App
+- There must be absolute isolation between installations of the App. E.g., customer 1 must not be able to access the resources of customer 2 via Roadie.
+
+# Routing the Callback URL Redirect
+
+GitHub Apps provide a single “Callback URL” configuration option. When this is set, the user installing the backstage app is redirected to the configured Callback URL.
+
+Because there is one Callback URL we had to create a single service shared by all customers that could be used to handle the installation redirect. So we did just that. We created a lambda API service that we called the “redirect-handler”. It would accept a “state” parameter that contains the URL target of the redirect. The redirect handler would simply unpack the state parameter and return the user a redirect to the target, reproducing all of the available url params contained in the request.
+
+![redirect-handler.png](redirect-handler.png)
+
+As you might be able to decern by, the state variable is what the redirect-handler uses to decide where to send the user on the redirect. This state is sent to the user by the backstage application, when the user indicates that they want to install the GitHub App. The state variable is a JWT token signed by a private key owned by the Backstage instance. The redirect-handler is configured with the required trust store to trust JWT tokens signed by the customer tenants. The JWT tokens contain a payload item called "urn:roadie:rdt".
+
+![app-installation-flow.png](app-installation-flow.png)
+
+When the redirect-handler recieves the request, it is decoding the JWT token in the state variable, verifies it is signed by a trusted source, extracts the "urn:roadie:rdt" variable and sends the user back a redirect to there.
+
+Now that the customers backstage instance has received the redirect, Backstage must verify that it trusts this installation.
+
+# Trusting the Installation ID returned via a redirect
+
+The customer backstage instance receives a request like the following, after the redirect hops have completed.
+
+```javascript
+GET - https://customern.roadie.so/github-installation?state=<jwt>&code=1234&installation_id=123456&setup_action=install
+```
+
+Roadie cannot just trust that the customer has provided an installation id that the user has ownership of. As such, Backstage must verify that the user is allowed to install this app. We have done this by enabling the feature "Request user authorization (OAuth) during installation" in GitHub Apps
+
+![request-user-authorization.png](request-user-authorization.png)
+
+With this setting, GitHub sends the redirect an addition field called "code". This field can be exchanged for a user access_token for the user that has installed the GitHub App. With this Backstage can look up the user's installations to make sure the user has access to that installation.
```

Comment:
"additional"

Id say remove the "for the user" and maybe write:
"This field can be exchanged for a user access_token that the Github App can use for authentication." or something.

"to make sure"-> "to verify"

Reply:
Ive totally rewritten.